### PR TITLE
Geocoding background importer

### PR DESCRIPTION
**Pull request by xavijam**
geocoding

New background geocoding polling available :dancer:.

Main features:
- Join import and geocoding proccesses in the same context, sharing 'background-polling' view.
- Geocoding jobs will be visible in the dashboard page. In dataset/map view, only datasets those belong to the visualization will be shown in the background-polling-view.
- Geocoding result dialogs included: success, error, no results,... etc.
- Geocoding already working dialog included.

---
background-geocoder
---
A silly diagram about new background polling view:
untitled drawing

Basically 'background-polling-view' listens what happens in the 'background-polling-model'. It any item is added, that view will create and add a new import/geocoding item in the list.

### CODE REVIEW

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4493/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4518/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4525/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4538/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4552/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4557/))

File: app/assets/stylesheets/common/background-polling/background-polling-result-details-dialog.css.scss [hunk 1]
```diff
@@ -5,52 +5,50 @@
...
   width: $w;
   margin: 0 auto;
   text-align: center;
+  border-top: 0;
 }
-.TwitterImportDetails-footerButton { display: inline-block }
-.TwitterImportDetails-footerButton span { line-height: 42px }
\ No newline at end of file
+.Dialog-footer .Button.BackgroundPollingDetails-footerButton { display: inline-block }
+.Dialog-footer .Button.BackgroundPollingDetails-footerButton span { line-height: 42px }
```
Comment (viddo) [inline app/assets/stylesheets/common/background-polling/background-polling-result-details-dialog.css.scss]: As I commented in earlier PRs, IMHO, `.Dialog-x.Something-else` would be better to have defined in the as `.Dialog-xxx--modifier` than to have them spread out in various places. Or as an alternative, define the (new?) component that is independent of `.Dialog-x` rules.

Comment (xavijam) [inline app/assets/stylesheets/common/background-polling/background-polling-result-details-dialog.css.scss]: I'd prefer a new component than add more styles in the "default" dialog stylesheet, becuase if not, it will be unmanageable in the near future.

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4578/))

File: lib/assets/javascripts/cartodb/common/background_polling/models/geocodings_collection.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  canGeocode: function() {
+    var ongoingGeocodings = 0;
+
+    this.each(function(m) {
+      if (!m.hasFailed() && !m.hasCompleted()) {
+        ++ongoingGeocodings
+      }
+    });
+
+    return ongoingGeocodings === 0;
```
Comment (viddo) [inline lib/assets/javascripts/cartodb/common/background_polling/models/geocodings_collection.js]: I think this method could be simplified:

1. Why not define the `ongoing` state on the individual model instead of here? I.e. implement a `m.isOngoing()` (which internally can check hasFailed & hasCompleted).
1. Since only caring if there's any ongoing or not, why not use [`_.any`](https://cdn.rawgit.com/jashkenas/underscore/1.4.4/index.html#some)? This way it doesn't have to iterate the whole list either. 

I.e.
```js
canGeocode: function() {
  return !this.any(function(m) {
    return m.isOngoing(); // or perhaps m.isProcessing() to make it clearer what is actually "ongoing"?
  });
}
```

Less and clearer, makes it easier to follow the intention w/o having to know internal details.

File: lib/assets/javascripts/cartodb/common/background_polling/views/geocodings/geocoding_result_details_view.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      state: this.model.get('state') || '',
+      realRows: realRows,
+      realRowsFormatted: Utils.formatNumber(realRows),
+      realRowsPluralize: pluralizeString('row', 'rows', realRows),
+      processedRows: processedRows,
+      processedRowsFormatted: Utils.formatNumber(processedRows),
+      processedRowsPluralize: pluralizeString('row', 'rows', processedRows),
+      processableRows: processableRows,
+      processableRowsFormatted: Utils.formatNumber(processableRows),
+      processableRowsPluralize: pluralizeString('row', 'rows', processableRows),
```
Comment (viddo) [inline lib/assets/javascripts/cartodb/common/background_polling/views/geocodings/geocoding_result_details_view.js]: Since repeated thrice, you could DRY this by a curried function:
```js
var rowOrRows = pluralizeString(this, 'row', 'rows'),
…
realRowsPluralize: rowOrRows(realRows),
processedRowsPluralize: rowOrRows(processedRows),
processableRowsPluralize: rowOrRows(processableRows),
```

**Comment by xavijam** [comment]
Sorry @viddo, so long review :(.

File: app/assets/stylesheets/common/background-polling/background-polling-result-details-dialog.css.scss (hunk 1 above)
Comment (viddo) [inline app/assets/stylesheets/common/background-polling/background-polling-result-details-dialog.css.scss]: Yeah, sure. I think this is related to the CSS organization I [commented on in my last PR](https://github.com/CartoDB/cartodb/pull/4192#issuecomment-114974787), I feel there are small variants in each context that forces us to add this kind of rules, we should try to eliminate them in favor having less rules and more consistent styles. Let's add it to the list to discuss next week :)

**Comment by viddo** [comment]
No worries

Overall :+1: added some comments for nitty picky details I saw.

File: lib/assets/javascripts/cartodb/table/table.js [hunk 4]
```diff
@@ -209,33 +224,49 @@ $(function() {
...
-          model: backgroundImporterModel,
+        this.backgroundPollingModel.bind('geocodingCompleted', function(mdl) {
+          // Refresh data in order to have cartodb_georef_status column updated
+          // but only if current layer data is the geocoded one
+          if (this.table && this.table.get('id') === mdl.get('table_name') && this.table.data) {
+            this.table.data().refresh()
+          }
+          // Refresh map
+          if (this.mapTab && this.mapTab.updateDataLayerView) {
+            this.mapTab.updateDataLayerView()
```
Comment (viddo) [inline lib/assets/javascripts/cartodb/table/table.js]: A curiosity, why do we need to do an explicit state change call here rather than having mapTab to listen to some event on its model to update itself?

Comment (xavijam) [inline lib/assets/javascripts/cartodb/table/table.js]: Good question. We have this way (requiring mapTab) or the way you say (listening those events in each "main" view), but It'd mean pass the ```backgroundPollingModel``` to each view, something we already do with the current ```geocoder``` variable. Not sure, what do you (@viddo @javisantana) recommend?

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4581/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4586/))

File: lib/assets/javascripts/cartodb/common/background_polling/models/geocodings_collection.js (hunk 2 above)
Comment (xavijam) [inline lib/assets/javascripts/cartodb/common/background_polling/models/geocodings_collection.js]: Agree :+1:

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4587/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4590/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4597/))
